tun/tun_illumos.go

```diff
@@ -0,0 +1,519 @@
+// }
+
+func (tun *NativeTun) Name() (string, error) {
+	return tun.name, nil
+}
+
+func (tun *NativeTun) Flush() error {
+        // TODO: can flushing be implemented by buffering and using sendmmsg?
+        return nil
+}
+
+func (tun *NativeTun) Read(buf []byte, offset int) (int, error) {
+	select {
+	case err := <-tun.errors:
+		return 0, err
+
+	default:
+		return tun.read_tun(buf[offset:])
+	}
+}
+
+func (tun *NativeTun) Write(buf []byte, offset int) (int, error) {
+	return tun.write_tun(buf[offset:])
+}
+
+func (tun *NativeTun) File() *os.File {
+	return tun.tunFile
+}
+
+func (tun *NativeTun) MTU() (int, error) {
+	return tun.mtu, nil
+}
+
+func (tun *NativeTun) Events() chan Event {
+	return tun.events
+}
+
+func (tun *NativeTun) Close() error {
+	if tun.ip_fd >= 0 {
+		ip_muxid, err := get_ip_muxid(tun.ip_fd, tun.name)
+		if err != nil {
+			return err
+		}
+
+		err = punlink(tun.ip_fd, ip_muxid)
+		if err != nil {
+			return err
+		}
+
+		unix.Close(tun.ip_fd)
+		tun.ip_fd = -1
+	}
+
+	if tun.tunFile != nil {
+		tun.tunFile.Close()
+		tun.tunFile = nil
+	}
+
+	if tun.events != nil {
+		close(tun.events)
+	}
+
+	return nil
+}
+
+func CreateTUNFromFile(file *os.File, mtu int) (Device, error) {
+	/*
+	 * XXX It's not currently clear to me how to take a TUN file descriptor
+	 * and determine the attached PPA.
+	 *
+	 * XXX Based on the current architecture of the daemon, we need to be
+	 * able to reconstitute our NativeTun object from just a file
+	 * descriptor number as the fd number is passed (with no other
+	 * information) through the environment to the child.  For now, use the
+	 * "-f" flag.
+	 */
+	return nil, fmt.Errorf("CreateTUNFromFile() not currently supported")
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	if name != "tun" {
+		return nil, fmt.Errorf("Interface name must be 'tun'")
+	}
+
+	/*
+	 * To establish a "tun" interface, we need to open a few file
+	 * descriptors.
+	 */
+	ip_node := "/dev/udp"
+	dev_node := "/dev/tun"
+
+	/*
+	 * First, the IP driver:
+	 */
+	ip_fd, err := unix.Open(ip_node, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("Could not open IP (%s)", ip_node)
+	}
+
+	/*
+	 * Now, the TUN driver.  Note that we use "os.OpenFile()" instead of
+	 * "unix.OpenFile()" so that we get the os.File object here.  The rest
+	 * of the Wireguard API seems to depend on that functionality.
+	 */
+	tunFile, err := os.OpenFile(dev_node, unix.O_RDWR, 0)
+	if err != nil {
+		unix.Close(ip_fd)
+		return nil, fmt.Errorf("Could not open TUN (%s)", dev_node)
+	}
+	fd := int(tunFile.Fd())
+
+	/*
+	 * Ask the TUN driver for a new PPA number:
+	 */
+	ppa, err := tun_new_ppa(fd)
+	if err != nil {
+		unix.Close(ip_fd)
+		tunFile.Close()
+		return nil, err
+	}
+
+	name = fmt.Sprintf("tun%d", ppa)
+	fmt.Printf("device %v\n", name)
+
+	/*
+	 * Open another temporary file descriptor to the TUN driver.
+	 * XXX It's not clear if this is actually needed, or if we could
```

Comment:
Unless there's surprising semantics in the Illumos API surface, seems like it should be fine to reuse the device FD you extracted from the os.File.